packages/attachments/src/model.ts

```diff
@@ -0,0 +1,456 @@
+   */
+  constructor(options: IAttachmentsModel.IOptions = {}) {
+    this.contentFactory = (options.contentFactory ||
+      AttachmentsModel.defaultContentFactory
+    );
+    this.map = new ObservableMap<IAttachmentModel>();
+    if (options.values) {
+      for (let key of Object.keys(options.values)) {
+        this._add(key, options.values[key]);
+      }
+    }
+    this.map.changed.connect(this._onMapChanged, this);
+
+    // If we are given a IModelDB, keep an up-to-date
+    // serialized copy of the AttachmentsModel in it.
+    if (options.modelDB) {
+      this._modelDB = options.modelDB;
+      this._serialized = this._modelDB.createValue('attachments');
+      if (this._serialized.get()) {
+        this.fromJSON(this._serialized.get() as nbformat.IAttachments);
+      } else {
+        this._serialized.set(this.toJSON());
+      }
+      this._serialized.changed.connect(this._onSerializedChanged, this);
+    }
+  }
+
+  /**
+   * A signal emitted when the model state changes.
+   */
+  get stateChanged(): ISignal<IAttachmentsModel, void> {
+    return this._stateChanged;
+  }
+
+  /**
+   * A signal emitted when the model changes.
+   */
+  get changed(): ISignal<this, IAttachmentsModel.ChangedArgs> {
+    return this._changed;
+  }
+
+  /**
+   * The keys of the attachments in the model.
+   */
+  get keys(): ReadonlyArray<string> {
+    return this.map.keys();
+  }
+
+  /**
+   * Get the length of the items in the model.
+   */
+  get length(): number {
+    return this.map ? Object.keys(this.map).length : 0;
+  }
+
+  /**
+   * The attachment content factory used by the model.
+   */
+  readonly contentFactory: IAttachmentsModel.IContentFactory;
+
+  /**
+   * Test whether the model is disposed.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  /**
+   * Dispose of the resources used by the model.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    this._isDisposed = true;
+    this.map.dispose();
+    Signal.clearData(this);
+  }
+
+  /**
+   * Whether the specified key is set.
+   */
+  has(key: string): boolean {
+    return this.map.has(key);
+  }
+
+  /**
+   * Get an item at the specified index.
+   */
+  get(key: string): IAttachmentModel {
+    return this.map.get(key);
+  }
+
+  /**
+   * Set the value at the specified index.
+   */
+  set(key: string, value: nbformat.IMimeBundle): void {
+    // Normalize stream data.
+    let item = this._createItem({ value });
+    this.map.set(key, item);
+  }
+
+  /**
+   * Clear all of the attachments.
+   */
+  clear(): void {
+    each(this.map.values(), (item: IAttachmentModel) => { item.dispose(); });
+    this.map.clear();
+  }
+
+  /**
+   * Deserialize the model from JSON.
+   *
+   * #### Notes
+   * This will clear any existing data.
+   */
+  fromJSON(values: nbformat.IAttachments) {
+    this.clear();
+    each(Object.keys(values), (key) => { this._add(key, values[key]); });
```

Comment:
This was copied straight from `outputarea/model`. I'll fix it here, but you might want to do another PR fixing this in outputarea.

Reply:
I don't mind doing that in a separate PR, but if you are doing it here and want to do it there, I think that's fine too, since they'll be in the same spot.